Clustering/Hdbscan/src/main/java/org/tribuo/clustering/hdbscan/HdbscanTrainer.java

```diff
@@ -0,0 +1,828 @@
+            Integer hierarchyLevel = entry.getKey();
+
+            int[] hierarchyLevelLabels = hierarchy.get(hierarchyLevel);
+
+            for (int i = 0; i < hierarchyLevelLabels.length; i++) {
+                int label = hierarchyLevelLabels[i];
+                if (clusterList.contains(label))
+                    clusterLabels.set(i, label);
+            }
+        }
+        return clusterLabels;
+    }
+
+    /**
+     * Produces the outlier score for each point in the data.
+     *
+     * @param clusters A list of Clusters forming a cluster tree which has already been propagated.
+     * @param pointNoiseLevels A double[] with the levels at which each point becomes noise.
+     * @param pointLastClusters An int[] with the last label of each point before becoming noise.
+     * @return A {@link DenseVector} containing the outlier scores.
+     */
+    private static DenseVector calculateOutlierScores(double[] pointNoiseLevels, int[] pointLastClusters,
+                                                                 List<HdbscanCluster> clusters) {
+
+        int numPoints = pointNoiseLevels.length;
+        DenseVector outlierScores = new DenseVector(numPoints);
+
+        // Iterate through each point, calculating its outlier score:
+        for (int i = 0; i < numPoints; i++) {
+            double epsilon_max = clusters.get(pointLastClusters[i]).getPropagatedLowestChildDeathLevel();
+            double epsilon = pointNoiseLevels[i];
+
+            double score = 0;
+            if (epsilon != 0)
+                score = 1-(epsilon_max/epsilon);
+
+            outlierScores.set(i, score);
+        }
+        return outlierScores;
+    }
+
+    /**
+     * Generate a map keyed by the cluster label and the data points assigned to them. The representation of a data
+     * points are their index in the dataset, and their outlier score. In this case, the data points are being sorted
+     * by their outlier scores, as convenience to quickly establish the exemplars from the dataset.
+     *
+     * @param clusterLabels A list containing every point's cluster label.
+     * @param outlierScoresVector A {@link DenseVector} containing the outlier scores.
+     * @return A map of the cluster labels, and the points assigned to them.
+     */
+    private static Map<Integer, TreeMap<Double, Integer>> generateClusterAssignments(List<Integer> clusterLabels,
+                                                                           DenseVector outlierScoresVector) {
+        // A map value of TreeMap<Double, Integer>> provides a sorting of the entries by outlier score.
+        Map<Integer, TreeMap<Double, Integer>> clusterAssignments = new HashMap<>();
+        for (int i = 0; i < clusterLabels.size(); i++) {
+            Integer clusterLabel = clusterLabels.get(i);
+            Double outlierScore = outlierScoresVector.get(i);
+            if (clusterAssignments.containsKey(clusterLabel)) {
```

Comment:
Can this be `computeIfAbsent`?

Reply:
I didn't see a nice way to do this with computeIfAbsent. If you have a code snippet, I'll take it though.

Reply:
I think this should be equivalent?

```
--- a/Clustering/Hdbscan/src/main/java/org/tribuo/clustering/hdbscan/HdbscanTrainer.java
+++ b/Clustering/Hdbscan/src/main/java/org/tribuo/clustering/hdbscan/HdbscanTrainer.java
@@ -665,14 +665,8 @@ final public class HdbscanTrainer implements Trainer<ClusterID> {
         for (int i = 0; i < clusterLabels.size(); i++) {
             Integer clusterLabel = clusterLabels.get(i);
             Double outlierScore = outlierScoresVector.get(i);
-            if (clusterAssignments.containsKey(clusterLabel)) {
-                clusterAssignments.get(clusterLabel).put(outlierScore, i);
-            }
-            else {
-                TreeMap<Double, Integer> outlierScoreIndexTree = new TreeMap<>();
-                outlierScoreIndexTree.put(outlierScore, i);
-                clusterAssignments.put(clusterLabel, outlierScoreIndexTree);
-            }
+            Map<Double, Integer> tm = clusterAssignments.computeIfAbsent(clusterLabel, (j) -> new TreeMap<>());
+            tm.put(outlierScore,i);
         }
         return clusterAssignments;
     }
```

Reply:
Yep, that works great, thanks